# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [86]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures
import numpy as np
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier


## 1. Loading the dataset

In [87]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

slime_data = pd.read_csv("science_data_large.csv")
print(slime_data.head(15))
slime_data.info()


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [88]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X = slime_data[['Temperature °C', 'Mols KCL']]
y = slime_data['Size nm^3']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [89]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
model = LinearRegression()
model.fit(X_train, y_train)

sample_data = pd.DataFrame({
    "Temperature °C": [25],
    "Mols KCL": [0.5]
})

prediction = model.predict(sample_data)
print("Prediction of the output of that sample with the trained model:", prediction)

score = model.score(X_test, y_test)
print("Model score:", score)

coeffs = model.coef_
intercept = model.intercept_
print("Coefficients:", coeffs)
print("Intercept:", intercept)

Prediction of the output of that sample with the trained model: [-387221.47171587]
Model score: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.479583408


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = 866.1464 \cdot x_1 + 1032.6951 \cdot x_2 - 409391.4796$

Report on score and explain meaning:

The default score function tells us how the independent variables, temp and mols, explain the how the target variable, size, varies. If the value is closer to 1, it shows that it is a better fit where the model explains all variance in the data while a value closer to 0 displays a model that doesn;t explain variability. 

## 4. Use Cross Validation

In [90]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance
from sklearn.model_selection import cross_val_score, KFold

kf = KFold(n_splits=5, random_state=42, shuffle=True)
scores = cross_val_score(model, X, y, cv=kf)
# print("Cross Validation scores for each fold:", scores)
# print("Average Cross Validation score:", scores.mean())

Write findings here:

Cross Validation scores for each fold: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average Cross Validation score: 0.8597294202684646

The average score of 0.86 tells us that th model explains about 86% of the variance in the target variable and hence fits the data well. 

The signicificance of these results is that the high scores allows us to realize the strength of the linear relationship between the features and the target variable. Also, the small differences between folds displays that it is consistent across different subsets. Therefore, this model would most likely perform well outside of the sample data.

## 5. Using Polynomial Regression

In [91]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

poly_model = LinearRegression()

kf = KFold(n_splits=5, random_state=42, shuffle=True)

scores_poly = cross_val_score(poly_model, X_train_poly, y_train, cv=kf)
print("Cross validation scores for each fold:", scores_poly)
print("Average cross validation score:", scores_poly.mean())

poly_model.fit(X_train_poly, y_train)

coeffs_poly = poly_model.coef_
intercept_poly = poly_model.intercept_
print("Polynomial Regression Coefficients:", coeffs_poly)
print("Polynomial Regression Intercept:", intercept_poly)


Cross validation scores for each fold: [1. 1. 1. 1. 1.]
Average cross validation score: 1.0
Polynomial Regression Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27199035e-07  1.26451322e-11
  2.00000000e+00  2.85714287e-02]
Polynomial Regression Intercept: 2.047925954684615e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$ h(x) = (2.0479 \times 10^{-5}) + (0)x + (12.0)x^2 + (-1.27199 \times 10^{-7})x^3 + (1.26451 \times 10^{-11})x^4 + (2.0)x^5 + (0.028571)x^6 $

Report on the score and interpret:

It achieved perfect cross validation scores meaning the model was able to explain 100% of the variance in the target variable, matching the observed values for each split. 
This tells us that the quadratic polynomial transformation is very effective at providing the relationship and also proving consistency and stability. But at the same time, perfect results like these are unlikely in the real world. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [92]:
# Load the dataset. Then train and evaluate the classification models.

ckd_data = pd.read_csv("ckd_feature_subset.csv")

X = ckd_data.drop("Target_ckd", axis=1)
y = ckd_data["Target_ckd"]

print(ckd_data.head())


        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1


## Training and evaluating the classification models

In [93]:
models = {
    "Logistic Regression": LogisticRegression(random_state=42, max_iter=1000),
    "SVC": SVC(random_state=42),
    "KNN": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(random_state=42, max_iter=1000)
}

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# use cross validation and store results
results = {}
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=skf, scoring='accuracy')
    results[name] = scores
    # print(f"{name} Accuracy scores: {scores}")
    # print(f"{name} Mean Accuracy: {scores.mean():.4f}, Std: {scores.std():.4f}\n")

# create summary table
results_table = pd.DataFrame({
    "Model": list(results.keys()),
    "Mean Accuracy": [np.mean(scores) for scores in results.values()],
    "Std Accuracy": [np.std(scores) for scores in results.values()]
})
print(results_table)


                 Model  Mean Accuracy  Std Accuracy
0  Logistic Regression       0.862366      0.033839
1                  SVC       0.933978      0.047300
2                  KNN       0.933978      0.047300
3       Neural Network       0.947527      0.039370


Based on the cross validation results, the neural network performed the best. It achieved a mean accuracy of approximately 94.75% with a standard deviation of 3.94%, which is higher than the other models. Logistic regression had a 86.24% mean accuracy, SVC had a 93.40% mean accuracy, and KNN had a 93.40% mean accuracy. The neural network is more effective at correctly classifying data and also more stable 

The higher mean accuracy shows that the neural network is more effective at correctly classifying the CKD data, and its relatively low standard deviation suggests that its performance is stable across different folds of the data.

## Results and Conclusion for Classification Experiments

In [94]:
configs = {
    "Single layer 50": MLPClassifier(random_state=42, max_iter=1000, hidden_layer_sizes=(50,)),
    "Single layer 100": MLPClassifier(random_state=42, max_iter=1000, hidden_layer_sizes=(100,)),
    "Two layers (100,50)": MLPClassifier(random_state=42, max_iter=1000, hidden_layer_sizes=(100, 50))
}

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

results = {}
for name, model in configs.items():
    scores = cross_val_score(model, X, y, cv=skf, scoring='accuracy')
    results[name] = scores
    # print(f"{name} Accuracy scores: {scores}")
    # print(f"{name} Mean Accuracy: {scores.mean():.4f}, Std: {scores.std():.4f}\n")

results_table = pd.DataFrame({
    "Configuration": list(results.keys()),
    "Mean accuracy": [np.mean(scores) for scores in results.values()],
    "Std accuracy": [np.std(scores) for scores in results.values()]
})
print(results_table)


         Configuration  Mean accuracy  Std accuracy
0      Single layer 50       0.940860      0.038374
1     Single layer 100       0.947527      0.039370
2  Two layers (100,50)       0.973763      0.024260


## Results and Conclusion for Neural Network Experiments

Looking at the single hidden layer to the two layer architecture, the two layer architecture resulted in a significant improvement in mean accuracy (94-95% to 97.38%). The standard deviation also dropped. We discover that the two layer models are more accurate and more consistent across different folds. 

Increasing the number of neurons in a single layer from 50 to 100 gave a slight improvment in performance but not as much as adding an additional layer. These results tell us that adding an extra hidden layer provides the most impact on improving classification performance. 

